pennylane/ops/op_math/linear_combination.py

```diff
@@ -0,0 +1,482 @@
+        ... )
+        >>> obs = qml.PauliZ(0) @ qml.PauliY(1)
+        >>> print(H.compare(obs))
+        True
+
+        >>> H1 = qml.LinearCombination([1, 1], [qml.PauliX(0), qml.PauliZ(1)])
+        >>> H2 = qml.LinearCombination([1, 1], [qml.PauliZ(0), qml.PauliX(1)])
+        >>> H1.compare(H2)
+        False
+
+        >>> ob1 = qml.LinearCombination([1], [qml.PauliX(0)])
+        >>> ob2 = qml.Hermitian(np.array([[0, 1], [1, 0]]), 0)
+        >>> ob1.compare(ob2)
+        False
+        """
+
+        if (pr1 := self.pauli_rep) is not None and (pr2 := other.pauli_rep) is not None:
+            pr1.simplify()
+            pr2.simplify()
+            return pr1 == pr2
+
+        if isinstance(other, (LinearCombination, qml.Hamiltonian)):
+            op1 = self.simplify()
+            op2 = other.simplify()
+            return op1._obs_data() == op2._obs_data()  # pylint: disable=protected-access
+
+        if isinstance(other, (Tensor, Observable)):
+            op1 = self.simplify()
+            return op1._obs_data() == {
+                (1, frozenset(other._obs_data()))  # pylint: disable=protected-access
+            }
+
+        raise ValueError(
+            "Can only compare a LinearCombination, and a LinearCombination/Observable/Tensor."
+        )
+
+    def __matmul__(self, other):
+        """The product operation between Operator objects."""
+        if isinstance(other, LinearCombination):
+            coeffs1 = copy(self.coeffs)
+            ops1 = self.ops.copy()
+            shared_wires = qml.wires.Wires.shared_wires([self.wires, other.wires])
+            if len(shared_wires) > 0:
+                raise ValueError(
+                    "Hamiltonians can only be multiplied together if they act on "
+                    "different sets of wires"
+                )
+
+            coeffs2 = other.coeffs
+            ops2 = other.ops
+
+            coeffs = qml.math.kron(coeffs1, coeffs2)
+            ops_list = itertools.product(ops1, ops2)
+            terms = [qml.prod(t[0], t[1], lazy=False) for t in ops_list]
+            return qml.LinearCombination(coeffs, terms)
+
+        if isinstance(other, Operator):
+            if other.arithmetic_depth == 0:
+                new_ops = [op @ other for op in self.ops]
+
+                # build new pauli rep using old pauli rep
+                if (pr1 := self.pauli_rep) is not None and (pr2 := other.pauli_rep) is not None:
+                    new_pr = pr1 @ pr2
+                else:
+                    new_pr = None
+                return LinearCombination(self.coeffs, new_ops, _pauli_rep=new_pr)
+            return qml.prod(self, other)
+
+        return NotImplemented
+
+    def __add__(self, H):
+        r"""The addition operation between a LinearCombination and a LinearCombination/Tensor/Observable."""
+        ops = self.ops.copy()
+        self_coeffs = copy(self.coeffs)
```

Comment:
Same question about the need for copies.

Reply:
The first copy is necessary in the current implementation as it gets extended below. All the others I can remove indeed :)